horovod/mxnet/__init__.py

```diff
@@ -87,3 +93,6 @@ def broadcast_parameters(params, root_rank=0):
     for _, p in params:
         broadcast_(p, root_rank, str(count))
         count += 1
+
+    # Make sure the tensors pushed to MXNet engine get processed
+    mx.nd.waitall()
```

Comment:
It seems waitall will prevent exceptions from being rethrown. Please see this: https://mxnet.apache.org/versions/master/architecture/exception_handling.html

Reply:
Then we need to put wait_to_read back in the loop

Reply:
Oh, I see.  What if you omit explicit op names and allow internals to generate operation IDs automatically (sequentially)?

Reply:
That can be an option but it might not be a good idea to apply here. This function is used to make sure each worker have the same set of parameters to start. We’d better make sure these ops gets executed such that each worker are synced before starting training.

Reply:
OK - is it possible to do the sync after the loop?  That way most of the operation would have a chance to finish before we start syncing the outputs.

Reply:
@yuxihu, does engine always process tensors sequentially, or it could do graph execution, in which case it may be a better idea to wait for all of them?

Reply:
The engine process tensors which have dependencies sequentially. For independent tensors, each of them is processed in a thread simultaneously. Given that params here are independent, it might be a good idea to wait all. Now we have two options:

1. waitall after the loop: this method looks cleaner but it suppresses exceptions.
2. wait_to_read for each tensor in the loop: not so efficient but have all we want: throwing exception, wait for all tensors. Since broadcast_parameters is one time action for each training, we might be able to tolerate the little inefficiency.

Please let us know what's your opinion on these two options.

Reply:
How about having two loops.  In the first loop, you enqueue all broadcast_ ops, and in the second loop you do all the waiting?

Reply:
Two loops sound good.